vortex-array/src/compute/is_sorted.rs

```diff
@@ -0,0 +1,154 @@
+use vortex_error::{VortexExpect, VortexResult, vortex_bail};
+
+use crate::arrays::{ConstantArray, NullArray};
+use crate::stats::{Precision, Stat};
+use crate::{Array, ArrayExt, Encoding};
+
+pub trait IsSortedFn<A> {
+    /// # Preconditions
+    /// - The array's length is > 1.
+    /// - The array is not encoded as `NullArray` or `ConstantArray`.
+    /// - If doing a `strict` check, if the array is nullable, it'll have at most 1 null element
+    ///   as the first item in the array.
+    fn is_sorted(&self, array: A) -> VortexResult<bool>;
+
+    fn is_strict_sorted(&self, array: A) -> VortexResult<bool>;
+}
+
+impl<E: Encoding> IsSortedFn<&dyn Array> for E
+where
+    E: for<'a> IsSortedFn<&'a E::Array>,
+{
+    fn is_sorted(&self, array: &dyn Array) -> VortexResult<bool> {
+        let array_ref = array
+            .as_any()
+            .downcast_ref::<E::Array>()
+            .vortex_expect("Failed to downcast array");
+        IsSortedFn::is_sorted(self, array_ref)
+    }
+
+    fn is_strict_sorted(&self, array: &dyn Array) -> VortexResult<bool> {
+        let array_ref = array
+            .as_any()
+            .downcast_ref::<E::Array>()
+            .vortex_expect("Failed to downcast array");
+        IsSortedFn::is_strict_sorted(self, array_ref)
+    }
+}
+
+#[allow(clippy::wrong_self_convention)]
+/// Helper methods to check sortedness with strictness
+pub trait IsSortedIteratorExt: Iterator
+where
+    <Self as Iterator>::Item: PartialOrd,
+{
+    fn is_strict_sorted(self) -> bool
+    where
+        Self: Sized,
+        Self::Item: PartialOrd,
+    {
+        self.is_sorted_by(|a, b| a < b)
+    }
+}
+
+impl<T> IsSortedIteratorExt for T
+where
+    T: Iterator + ?Sized,
+    T::Item: PartialOrd,
+{
+}
+
+pub fn is_sorted(array: &dyn Array) -> VortexResult<bool> {
+    is_sorted_opts(array, false)
+}
+pub fn is_strict_sorted(array: &dyn Array) -> VortexResult<bool> {
+    is_sorted_opts(array, true)
+}
+
+fn is_sorted_opts(array: &dyn Array, strict: bool) -> VortexResult<bool> {
```

Comment:
Maybe it's worth unrolling this into two functions

Reply:
pushed something, I do think it looks better now.